One test case to rule them all

jdavcs opened the pull request:
This is based on #8737 (one new commit).

This adds integration tests for all configuration properties specified in the schema. Out of these, 226 pass locally (tested on 2 machines, so I hope the results will be the same on Jenkins); another 56 are commented out because they require edits to code outside of this file (I will address them in a follow-up). Added time: 18 sec (when running locally). [updated: 229->226/53->56: see last commit]

This commit does not change anything in the code other than adding this file. Its purpose is to establish a ground truth prior to any further refactoring of the configuration system (and, thus, catch any unintended side effects during subsequent edits). Together with the recently added unit tests, this should make modifying configuration-related code a little safer, I hope.

WHAT IS TESTED
This tests (1) automatic creation of configuration properties; and (2) assignment of default values that are specified in the schema and, in some cases, are also processed at load time (paths resolved, csv strings converted to lists, etc). 
For example:
- for property `foo` defined in the schema with a default value of `42`, the test will ensure that `app.config.foo == 42`
- if `foo`'s default value is further processed at load time (e.g., listified), the test would ensure that
`app.config.foo == [42]`, (or `== [42, 43, 44]` if the default was a csv str `42,43,44`
- if `foo`'s default was not set, the test ensures its value is null
- if `foo`'s default was not set, but a default value was assigned at load time, the test will test that value. And so forth. [for more details, see comment at the top of the new file]

I based this on top of my previous PR (as opposed to `dev`) because some of the follow-up commits will use the new `path_resolve_to` feature from #8689. And even if they didn't, basing this on `dev` would lead to 2 separate branches, each with many edits to the same set of files.

jdavcs committed:
Add integration tests for all config options

This tests: (1) automatic creation of configuration properties; and
(2) assignment of default values that are specified in the schema and, in
some cases, are also processed at load time (paths resolved, csv strings
converted to lists, etc). [see comment at the top of the new file]

229 pass
53 commented out and require code editing and will be addressed in
subsequent commites.

jdavcs commented:
Rebased to include bd9f188 which fell through the cracks in a series of rebases.

jdavcs commented:
Rebased to exclude 3 tests that test values that are overridden for testing on non-sqlite dbs. Locally they pass of course; but not on Jenkins.

mvdbeek commented on test/integration/test_config_defaults.py:
```diff
@@ -0,0 +1,1107 @@
+    # def test_default_database_engine_option_max_overflow(self):
+    # TODO: default value overridden for tests runnign on non-sqlite databases
+
+    def test_default_database_engine_option_pool_recycle(self):
```
Could you parametrize all the `test_.*` methods using @pytest.mark.parametrize ?
I'd build a list of config strings up front (or even discover them from the schema) and then parametrize a generic function, i.e
```python
CONFIG_PROPERTIES = [
    'database_engine_option_server_side_cursors',
    'database_engine_option_pool_recycle',
]


@pytest.mark.parametrize('config_property', CONFIG_PROPERTIES)
def test_default_config_property(config_property):
    expect = self.get_default(config_property)
    assert expect == getattr(self._app.config, config_property)
```

If you can auto-discover these from the schema: even better :)

jdavcs replied:
Probably yes. Marking this as WIP for now, while I'm consolidating tests.

jdavcs replied:
@mvdbeek how's this? I can barely stop myself from adding 'one test to find them' to the commit message :)

jdavcs committed:
Replace all tests with a parametrized generic test

jdavcs reviewed (commented):
The new commit gets rid of 915 lines of code. Thank you for the suggestion, @mvdbeek! 
The approach is the same as here: https://github.com/galaxyproject/galaxy/commit/9e07d9cc062bf8a21bfab72c2505191e1d3c86e4#diff-cb4f71ec7457ece91d85d328b489f3da

I've tried to make this as least hackish as possible. The benefits, I think, outweigh the relative ugliness: 227 tests got reduced to just one simple parametrized test. Also, if an option (or a group of options) require a special way to calculate their expected value, that's done by simply defining a function, and adding a new item to the `CUSTOM` dictionary. Of course, a regular test can be added too: as long as it's added after the generic test, it can use `DRIVER` to access the config object.

jdavcs commented:
I just realized this overrides a revert in my previous merged PR https://github.com/galaxyproject/galaxy/commit/b962807c8214ee07e82b2b4e6e2c4219081f26e4
Fixed and rebased.

nsoranzo commented:
> The new commit gets rid of 915 lines of code. Thank you for the suggestion, @mvdbeek!

Maybe it could make sense to squash (some of) the commits.

jdavcs commented:
@nsoranzo Yes, will do. Apparently, I got 4 extra commits in during a rebase (not my commits). 
And done.

mvdbeek commented on test/integration/conftest.py:
```diff
@@ -0,0 +1,8 @@
+def pytest_unconfigure(config):
+    try:
+        # This needs to be run if no test were run.
```
That doesn't look quite right but is probably necessary because these tests are a mix of unittest and pytest. Normally all fixtures should cleanup after themselves.

jdavcs replied:
We do need it, I think (and this one should be pure pytest): if you keep that test, but don't pass any data, then the fixture will be called, but no tests will be generated, and the unconfigure hook will clean up. Remove the hook, and the fixture will keep running until you kill it manually.

mvdbeek commented on test/integration/test_config_defaults.py:
```diff
@@ -1,22 +1,191 @@
+
+# TODO: split into (1) do not test; and (2) todo: fix and test
+DO_NOT_TEST = [
+    'config_dir',  # value overridden for testing
```
So now that we have a nice getter/setter system we could probably also record when something is not the default value.
Not for this PR of course.

jdavcs replied:
Yes, I'm sure we can! And I'm certainly not done with this -- but yes, that would be for a subsequent PR. (as well as cleaning up/normalizing the getter setter system)

mvdbeek reviewed (approved):
Very nice! Ready to merge @ic4f ?

jdavcs commented:
> Very nice! Ready to merge @ic4f ?

Yes!
